system_tests/conditionaltx_test.go

```diff
@@ -0,0 +1,350 @@
+	currentRootHash2 := getStorageRootHash(t, node, contractAddress2)
+	currentSlotValueMap2 := getStorageSlotValue(t, node, contractAddress2)
+
+	rpcClient, err := node.Stack.Attach()
+	testhelpers.RequireImpl(t, err)
+
+	l2info.GenerateAccount("User2")
+
+	testConditionalTxThatShouldSucceed(t, ctx, -1, l2info, rpcClient, nil)
+
+	block, err := l1client.BlockByNumber(ctx, nil)
+	testhelpers.RequireImpl(t, err)
+	blockNumber := block.NumberU64()
+	blockTime := block.Time()
+	successOptions := getSuccessOptions(contractAddress1, contractAddress2, currentRootHash1, currentRootHash2, currentSlotValueMap1, currentSlotValueMap2, blockNumber, blockTime)
+	for i, options := range successOptions {
+		testConditionalTxThatShouldSucceed(t, ctx, i, l2info, rpcClient, options)
+	}
+
+	tx, err = simple1.Increment(&auth)
+	Require(t, err, "failed to call Increment()")
+	_, err = EnsureTxSucceeded(ctx, l2client, tx)
+	Require(t, err)
+	tx, err = simple2.Increment(&auth)
+	Require(t, err, "failed to call Increment()")
+	_, err = EnsureTxSucceeded(ctx, l2client, tx)
+	Require(t, err)
+
+	previousStorageRootHash1 := currentRootHash1
+	currentRootHash1 = getStorageRootHash(t, node, contractAddress1)
+	if bytes.Equal(previousStorageRootHash1.Bytes(), currentRootHash1.Bytes()) {
+		testhelpers.FailImpl(t, "storage root hash didn't change as expected")
+	}
+	previousSlotValueMap1 := currentSlotValueMap1
+	currentSlotValueMap1 = getStorageSlotValue(t, node, contractAddress1)
+
+	previousStorageRootHash2 := currentRootHash2
+	currentRootHash2 = getStorageRootHash(t, node, contractAddress2)
+	if bytes.Equal(previousStorageRootHash2.Bytes(), currentRootHash2.Bytes()) {
+		testhelpers.FailImpl(t, "storage root hash didn't change as expected")
+	}
+	previousSlotValueMap2 := currentSlotValueMap2
+	currentSlotValueMap2 = getStorageSlotValue(t, node, contractAddress2)
+
+	block, err = l1client.BlockByNumber(ctx, nil)
+	testhelpers.RequireImpl(t, err)
+	blockNumber = block.NumberU64()
+	blockTime = block.Time()
+	successOptions = getSuccessOptions(contractAddress1, contractAddress2, currentRootHash1, currentRootHash2, currentSlotValueMap1, currentSlotValueMap2, blockNumber, blockTime)
+	for i, options := range successOptions {
+		testConditionalTxThatShouldSucceed(t, ctx, i, l2info, rpcClient, options)
+	}
+	block, err = l1client.BlockByNumber(ctx, nil)
+	testhelpers.RequireImpl(t, err)
+	blockNumber = block.NumberU64()
+	blockTime = block.Time()
+	future := hexutil.Uint64(blockTime + 30)
+	past := hexutil.Uint64(blockTime - 1)
+	futureBlockNumber := hexutil.Uint64(blockNumber + 1000)
+	currentBlockNumber := hexutil.Uint64(blockNumber)
+	if blockNumber == 0 {
+		testhelpers.FailImpl(t, "internal test error: unexpected blockNumber == 0")
+	}
+	previousBlockNumber := hexutil.Uint64(blockNumber - 1)
+	failOptions := []*arbitrum_types.ConditionalOptions{
+		{KnownAccounts: map[common.Address]arbitrum_types.RootHashOrSlots{contractAddress1: {RootHash: &previousStorageRootHash1}}},
```

Comment:
I've added a simple `optionsProduct` operation to the tests. It probably is possible to figure out better rules for combining two `ConditionalOptions` objects to avoid having to de-duplicate results when using it for creating test cases, but I am not sure if that's necessary. If it is worth spending more time, or if you have some idea how to simplify it I will be happy to continue on it, but I believe that it should be a good enough for unit test purposes. Please let me know what you think :)